mysql/mysql_resource_client.go

```diff
@@ -0,0 +1,122 @@
+package mysql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/mitchellh/mapstructure"
+	"net/http"
+)
+
+// ResourceClient is an AuthenticatedClient with some additional information about the resources to be addressed.
+type ResourceClient struct {
+	*MySQLClient
+	ResourceDescription string
+	ContainerPath       string
+	ResourceRootPath    string
+	ServiceInstanceID   string
+}
+
+func (c *ResourceClient) createServiceInstanceResource(requestBody interface{}, responseBody interface{}) error {
+
+	var objectPath = c.getContainerPath(c.ContainerPath)
+	c.client.DebugLogString(fmt.Sprintf("[Debug] : Trying to create ServiceInstance at %s", objectPath))
+	_, err := c.executeRequest("POST", objectPath, requestBody)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ResourceClient) getServiceInstanceResource(instanceName string, responseBody interface{}) error {
+
+	var objectPath = c.getObjectPath(c.ResourceRootPath, instanceName)
+
+	resp, err := c.executeRequest("GET", objectPath, nil)
+	if err != nil {
+		return err
+	}
+
+	return c.unmarshalResponseBody(resp, responseBody)
+
+}
+
+// ServiceInstance needs a PUT and a body to be destroyed
+func (c *ResourceClient) deleteServiceInstanceResource(name string, requestBody interface{}) error {
+	var objectPath string
+	if name != "" {
+		objectPath = c.getObjectPath(c.ResourceRootPath, name)
+	} else {
+		objectPath = c.ResourceRootPath
+	}
+	_, err := c.executeRequest("PUT", objectPath, requestBody)
+	if err != nil {
+		return err
+	}
+
+	// No errors and no response body to write
+	return nil
+}
+
+func (c *ResourceClient) createAccessRuleResource(requestBody interface{}, responseBody interface{}) error {
```

Comment:
Are we going to be adding more resources than just ServiceInstances and AccessRules? It looks like the all but the ServiceInstance API calls need a Service InstanceID so if we are going to add more of the resources we should let's the AccessRule methods here.

Reply:
I don't think we will be adding any other types of resources.
I totally didn't get what you mean about the AccessRule methods. Could you please elaborate a bit more ?

Reply:
Oh wow. I don't even know what I wrote. My apologies. The `mysql_resource_client` file should be home to the generic methods that each resource calls (ie. createResource, updateResource, etc.). Right now, this file has two specific resource calls (instance and access rule), and these are needed because of the way the api is written where everything but an instance needs the instance id in their uri. We can keep it like this if there are no plans to add additional resources but we'll have to make this file more generic if that ever changes.

Reply:
ok. I don't think that there will be any more resources being added. I spent some time thinking about this and I agree with you that it should be put into 2 different resource files to keep it clean. I'll refactor to 2 files so we don't have to deal with it again later in the future, as well as to maintain some kind of consistency.

Reply:
I've refactored the access_rule methods into access_rule_resource_client.

Fixed.